lib/libc/common/Kconfig

```diff
@@ -12,14 +12,42 @@ config COMMON_LIBC_MALLOC
 	  Common implementation of malloc family that uses the kernel heap
 	  API.
 
+config COMMON_LIBC_MALLOC_ARENA_SIZE_DEFAULT
+	int
+	default 0 if MINIMAL_LIBC
+	default 16384 if MMU
+	default -1
+	help
+	  Indicates the default value for COMMON_LIBC_MALLOC_ARENA_SIZE.
+	  This is used when COMMON_LIBC_MALLOC_ARENA_SIZE is set to -2.
+
+	  When using the minimal C library, the default is to have no
+	  malloc heap. Otherwise, on systems with an MMU the default is
+	  16kB and all other systems will default to using all remaining
+	  ram for the malloc heap.
+
 config COMMON_LIBC_MALLOC_ARENA_SIZE
 	int "Size of the common C library malloc arena"
 	depends on COMMON_LIBC_MALLOC
-	default 0
+	default -2
 	help
 	  Indicate the size in bytes of the memory arena used for
 	  common C library malloc() implementation.
 
+	  If set to -2, then the value of
+	  COMMON_LIBC_MALLOC_ARENA_SIZE_DEFAULT will be used instead, and
+	  the remaining rules applied.
+
+	  If set to zero, then no malloc() heap will be available.
+
+	  If set to -1, then all remaining system RAM will be used for this
+	  area.
+
+	  If user mode is enabled, and MPU hardware has requirements that
+	  regions be sized to a power of two and aligned to their size,
+	  then this must be defined as a power of two or a compile error
+	  will result.
```

Comment:
Having both `COMMON_LIBC_MALLOC_ARENA_SIZE` and `COMMON_LIBC_MALLOC_ARENA_SIZE_DEFAULT` is quite confusing and seems far from ideal to me.

Instead of making `COMMON_LIBC_MALLOC_ARENA_SIZE` function as both the "master switch" for malloc as well as the size of the malloc arena, why not make it specify only the size of the malloc arena, and instead make `COMMON_LIBC_MALLOC` the master switch as it should be?

Reply:
yeah, as usual it's complicated by the desire to remain compatible with existing application configurations that set `MINIMAL_LIBC_MALLOC_ARENA_SIZE` (or `PICOLIBC_HEAP_SIZE`, once that patch is merged). 

Hrm. I bet this would look better if I changed the default value of the legacy configuration values from `0` to `-1` so I could tell the difference between an application which explicitly defines the heap size as the old default and one which doesn't define the value at all.

I'll take a stab at that and see what it looks like.

Reply:
To answer the question about "master switch", `malloc` et al. will be available whenever `COMMON_LIBC_MALLOC` is defined, but if the heap is set to zero, then it will never succeed. This makes it work like the minimal C library allocator, but it seems like a user trap to me. Your application would build and run until the first malloc call, at which point it would fail. Should we fix that?

Reply:
in any case, moving the funky bits for dealing with legacy settings out of the common malloc settings and into the legacy library settings makes things look a lot cleaner. It does mean that the legacy settings will override the common settings, I think that's probably fine, but would appreciate some thought on that.